langkit/compiled_types.py

```diff
@@ -2404,6 +2419,42 @@ def __init__(self, name, location, doc, base, fields,
         entity_type = self.entity
         del entity_type
 
+    @property
+    def doc(self):
+        result = super(ASTNodeType, self).doc
+
+        # If this is a list node and that parsers build it, add a precise list
+        # of types it can contain: the element type might be too generic.
+        if self.is_list and not self.synthetic:
+            precise_types = (self.precise_list_element_types
+                             .minimal_matched_types)
+            if len(precise_types) > 1:
+                result_list = result.splitlines() if result else []
+                type_descr = [
+                    'This list node can contain one of the following nodes:'
+                ] + list(sorted('* {}'.format(t.dsl_name)
+                                for t in precise_types))
+
+                # Indent lines in `type_descr` as much as the first non-null
```

Comment:
Wouldn't a bit of mako have helped there ? this is particularly hard to read :/

Reply:
As discussed on Slack: unfortunately the result is piped to Sphinx, which is particularly picky about formatting aspects such as indentation. Using Mako in this context would probably make things worse. That being said, I could probably improve readability of this moving text formatting primitives to `langkit.utils` for instance. I’ll see what I can do.